### DISCUSSION TITLE
weird behaviour with checkboxes

### DISCUSSION THREAD

**Discussion by talshadar** [livewire/livewire#8412] [1. Help]
Updated to include code.

So first off, I did find a solution.  That being said, it is quite awkward compared to how I've done using older versions of laravel or livewire.  So this is mostly asking whether I'm missing something - or if not, why was changes like this made? Because to be honest, it makes absolutely no sense.

I'm building a page to allow the user to select properties to associate with an item.  Items are in 1 table, properties in another, with a relationship table in between holding the item_id and property_id.

What I want:
I get all the properties - then do a foreach loop creating the checkboxes using the property ids.. They select the ones they want, hit save - it sends back the ids of the selected properties, it then loops and saves them to the relationship table.  All good.  And it's exactly how I've done it in the past.

But with livewire 3 I have to do something completely different.

I do the foreach loop BUT instead of using the property Ids, I have to create/use an interger to hold a counter the increments by 1 for each property. And it doesn't send back the ids (the values) - it only sends back true or false.  (or empty in between the checkboxes: if you've selected 1 and 5 it returns: true, , , , false.  Which I resolved by pre-filling the array in the form with false.  then it's all false values except for what I selected.  Then I have to loop through selected properties - again with an incremental counter - and grab the actual property id from an array that holds those - than I can finally save it in the relationship table.

Bizarre behaviour is that if I used the ids for the properties that went say 1,2,3,4,6 (because sometimes things get deleted) - it would want to send back an array with 7 elements from 0 to 6.  or 0 - 10 etc

I saw a discussion saying to use wire:click to set the value but that didn't work at all.

My solution works - it just seems overly complicated and it doesn't make sense to me to change the behaviour of how checkboxes work that way.  Not returning the value is odd enough, but having it create empty elements in an array like that is unexpected behaviour, and something that doesn't make sense to me.

I didn't bother posting code (though I can if asked) because, like I said, I solved it. This is more about a discussion where (hopefully!!) someone can point out where/how I did it wrong or misunderstood it's behaviour.

Sorry, I saw someone mention how to differentiate between HTML and PHP but I couldn't remember. This is my first post.

CODE:
Blade file Component  that uses a form
```blade
                <div class="col-span-3 flex grid grid-flow-row-dense grid-cols-6 gap-2">
                <div class="col-span-6"><h2>{{ __('Properties') }}</h2></div>
                    @foreach ($properties as $property)
                        <div >
                            <input wire:model.defer="form.selectedProperties.{{ $propertiesCounter }}" type="checkbox"
                            value="{{ $propertiesCounter }}"  id="{{$property->name}}" class="" >
                                        {{$property->name}} &nbsp; 
                        </div>
                        @php $propertiesCounter++ @endphp
                    @endforeach
                </div>    
```

Form php 

```php
        $i = 0;
        foreach ( $this->selectedProperties as $value)
        {
            if ( $value == true)
            {
                HerbProperties::create(['herb_id' => $herb->id, 'property_id' => $this->allProperties[$i]->id  ]);
            }
            $i++;
        }
````

To set the properties to true when editing/viewing 
```php
        $pI = 0;
        foreach ($this->allProperties as $property)
        {
            if( $this->herb[0]->herbProperties->contains('property_id', $property->id) )
            {
                $this->selectedProperties[ $pI ] = true; 
            }
            $pI++;
        }
```

The first issue is that the checkboxes only send True or False - not the actual value. 
The second issue is the way it fills the numbers for the check boxes.

If I used just the property id instead of the propertyCounter in the blade wire:model it would add numbers for any missing gaps.

If I had property_ids of 1,2,4,5,7 and select 1,4,7 - instead of returning true,false,true,false,true - it returns true,false,false,true,false,false,true.

As for how I expected it to work:

1) Replace the propertyCounter with property_id so I'd know which property id had been selected vs having to do loops to try to match it up. Maybe other people haven't come across this particular behaviour because their records don't have gaps in them - or they found a better way to do it.

The sending true/false - that's workable.

As a side note, once I did build the correct solution - wow - it works so smoothly - absolutely love how quick and responsive it is. Extremely happy with how it's all working together.

**Comment by joshhanley** [comment]
@talshadar yeah posting code would help as I'm struggling to understand what you mean. If you can show the difference between V2 and V3 maybe there is a bug there we can look into.

**Reply by talshadar** [reply]
I updated my post with code. I don't have any with livewire V2 - my reference to that was purely based on everything I was finding regarding the changed behaviour of V3.

**Reply by joshhanley** [reply]
@talshadar thanks! Ah yep no wories. The formatting is weird, any chance you could fix that? 

When you paste code into github, you need to surround it in triple back ticks and then specify the file type like php or blade.

For example
\```php
\\ PHP code here
\```

\```blade
\\ HTML and blade code here
\```

**Reply by talshadar** [reply]
thank you! fixed the formatting.

**Reply by joshhanley** [reply]
@talshadar thanks! So is that the code for your current work around? If so, then yeah that isn't not ideal needing the counter.

Could you share the code for how you'd prefer/ expect it to be written (without the counter), so we can see what that's not working properly?

**Reply by talshadar** [reply]
I'll update with that - but also the work around I had to do to get the properties to show true if selected.
